src/API.jl

```diff
@@ -0,0 +1,557 @@
+        end
+        return pkgs
+    else
+        BrewPkg[]
+    end
+end
+
+"""
+outdated()
+
+Returns a list of all installed packages that are out of date as a `Vector{BrewPkg}`
+"""
+function outdated()
+    json_str = brewchomp(`outdated --json=v1`)
+    if length(json_str) == 0
+        return BrewPkg[]
+    end
+    brew_outdated = JSON.parse(json_str)
+
+    outdated_pkgs = BrewPkg[info(pkg["name"]) for pkg in brew_outdated]
+    return outdated_pkgs
+end
+
+"""
+refresh(;verbose=false)
+
+Forcibly remove all packages and add them again.  This should only be used to
+fix a broken installation, normal operation should never need to use this.
+"""
+function refresh!(;verbose=false)
+    pkg_list = list()
+    for pkg in pkg_list
+        rm(pkg,verbose=verbose)
+    end
+    for pkg in pkg_list
+        add(pkg,verbose=verbose)
+    end
+end
+
+"""
+upgrade()
+
+Iterate over all packages returned from `outdated()`, removing the old version
+and adding a new one.  Note that we do not simply call `brew upgrade` here, as
+we have special logic inside of `add()` to install from our tap before trying to
+install from mainline Homebrew.
+"""
+function upgrade()
+    # We have to manually upgrade each package, as `brew upgrade` will pull from mxcl/master
+    for pkg in outdated()
+        rm(pkg)
+        add(pkg)
+    end
+end
+
+"""
+json(names::Vector{AbstractString})
+
+For each package name in `names`, return the full JSON object for `name`, the
+result of `brew info --json=v1 \$name`, stored in a dictionary keyed by the names
+passed into this function. If `brew info` fails, throws an error. If `brew info`
+returns an empty object "[]", that object is represented by an empty dictionary.
+
+Note that running `brew info --json=v1` is somewhat expensive, so we cache the
+results in a global dictionary, and batching larger requests with this function
+similarly increases performance.
+"""
+const json_cache = Dict{String,Dict{AbstractString,Any}}()
+function json{T<:AbstractString}(names::Vector{T})
+    # First, normalize all names
+    names = String[normalize_name(n) for n in names]
+
+    # This is the dictionary of responses we'll return
+    objs = Dict{String,Dict{AbstractString,Any}}()
+
+    # Build list of names we have to ask for, eschewing asking for caching when we can
+    ask_names = String[]
+    for name in names
+        if haskey(json_cache, name)
+            objs[name] = json_cache[name]
+        else
+            push!(ask_names, name)
+        end
+    end
+
+    # Now ask for all these names if we have any
+    if length(ask_names) > 0
+        try
+            jdata = JSON.parse(brewchomp(Cmd(String["info", "--json=v1", ask_names...])))
+            for idx in 1:length(jdata)
+                json_cache[ask_names[idx]] = jdata[idx]
+                objs[ask_names[idx]] = jdata[idx]
+            end
+        catch
+            rethrow()
+            throw(ArgumentError("`brew info` failed for $(name)!"))
```

Comment:
Ah, good catch.  The `rethow()` was a holdover from debugging something inside the loop.